SpatialGDK/Source/SpatialGDK/Private/Interop/SkeletonEntities.cpp

```diff
@@ -0,0 +1,485 @@
+
+			check(EntityPtr != nullptr);
+
+			ConsiderEntityPopulation(EntityId, *EntityPtr);
+		});
+
+		if (OnManifestUpdated)
+		{
+			OnManifestUpdated(*ManifestEntityId, *Manifest);
+		}
+		else
+		{
+			GetCoordinator().SendComponentUpdate(*ManifestEntityId, Manifest->CreateComponentUpdate(), {});
+		}
+
+		if (Manifest->PopulatedEntities.Num() == Manifest->EntitiesToPopulate.Num())
+		{
+			UE_LOG(LogSpatialSkeletonEntityCreator, Log, TEXT("All %d entities populated"), Manifest->PopulatedEntities.Num());
+			Stage = EStage::SigningManifest;
+		}
+	}
+
+	if (Stage == EStage::SigningManifest)
+	{
+		// All locally populated entities need entity refreshes as this worker won't receive
+		// deltas for all the components it has added, and these can impact entity completeness.
+		for (const Worker_EntityId EntityId : Manifest->PopulatedEntities)
+		{
+			GetCoordinator().RefreshEntityCompleteness(EntityId);
+		}
+
+		Stage = EStage::Finished;
+	}
+}
+
+void FSkeletonEntityPopulator::PopulateEntity(Worker_EntityId SkeletonEntityId, AActor& SkeletonEntityStartupActor)
+{
+	EntityFactory Factory(NetDriver, NetDriver->PackageMap, NetDriver->ClassInfoManager, NetDriver->RPCService.Get());
+
+	NetDriver->PackageMap->ResolveEntityActor(&SkeletonEntityStartupActor, SkeletonEntityId);
+
+	TArray<FWorkerComponentData> Components;
+
+	// HACK: EntityFactory::WriteUnrealComponents expects an UnrealMetadata to be present in the components array,
+	// but since we aren't creating a complete entities but only fleshing out an existing one, we have to get this
+	// specific component from the entity in view.
+	Components.Emplace(NetDriver->Connection->GetCoordinator()
+						   .GetView()[SkeletonEntityId]
+						   .Components.FindByPredicate(ComponentIdEquality{ UnrealMetadata::ComponentId })
+						   ->GetWorkerComponentData());
+
+	USpatialActorChannel* Channel = NetDriver->ActorSystem->SetUpActorChannel(&SkeletonEntityStartupActor, SkeletonEntityId);
+
+	check(IsValid(Channel));
+
+	uint32 BytesWrittenDummy = 0;
+	Factory.WriteUnrealComponents(Components, Channel, BytesWrittenDummy);
+	Components.Add(ActorSystem::CreateLevelComponentData(SkeletonEntityStartupActor, *NetDriver->GetWorld(), *NetDriver->ClassInfoManager));
+
+	// HACK: Remove UnrealMetadata as we don't want to update it (see above),
+	// AND it's already in the View's components array.
+	Components.RemoveAll([](const FWorkerComponentData& Component) {
+		return Component.component_id == UnrealMetadata::ComponentId;
+	});
+
+	TArray<ComponentData> OwningComponent;
+
+	Algo::Transform(Components, OwningComponent, &Convert);
+
+	UE_LOG(LogSpatialSkeletonEntityCreator, Log, TEXT("EntityId: %lld Fleshing out with %d components"), SkeletonEntityId,
+		   OwningComponent.Num());
+
+	ViewCoordinator& Coordinator = NetDriver->Connection->GetCoordinator();
+
+	for (ComponentData& ComponentData : OwningComponent)
+	{
+		Coordinator.SendAddComponent(SkeletonEntityId, MoveTemp(ComponentData), {});
+	}
```

Comment:
Check if we can drop most of this in favor of creating the channel and UnrealComponents during normal replication.

Reply:
I think we need to be careful about this. We shouldn't be doing normal replication at this point as we want the entities to all exist and be filled before we start ticking

Reply:
Discussed this with @samiwh and @m-samiec offline; the conclusion was to try this out as there doesn't seem to be a requirement to always populate skeletons before the game can begin.

Reply:
Seems to work correctly.